rx_signal_processing/dsp.cu

```diff
@@ -133,17 +158,65 @@ namespace {
                                                           Eigen::Dynamic,
                                                           Eigen::Dynamic, 
                                                           Eigen::RowMajor>>(phases_cast, 
-                                                                beam_direction_counts[beamdir_num], 
-                                                                num_antennas);
+                                                                            num_beams, 
+                                                                            num_antennas);
 
-      auto result = phases * samps;
+      // Result matrix has dimensions beams x num_samples. This means one set of samples for 
+      // each beam dir. Eigen overloads the * operator so we dont need to implement any matrix
+      // work ourselves.
+      auto result = phases * samps;  
       
-      auto beamformed_cast = reinterpret_cast<std::complex<float>*>(beamformed_samples.data());
+      // This piece of code just transforms the Eigen result back into our flat vector.
+      auto beamformed_cast = reinterpret_cast<std::complex<float>*>(result_ptr);
       Eigen::Map<Eigen::Matrix<std::complex<float>, Eigen::Dynamic, 
                                 Eigen::Dynamic, Eigen::RowMajor>>(beamformed_cast, result.rows(), 
                                                                   result.cols()) = result;
+    };
+
+    auto main_phase_offset = 0;
+    auto main_results_offset = 0;
+
+    // Now we calculate the offsets into the samples, phases, and results vector for each
+    // RX frequency. Each RX frequency could have a different number of beams, so we increment
+    // the phase and results offsets based off the accumulated number of beams. Once we have the
+    // offsets, we can call the beamforming lambda.
+    for (uint32_t rx_freq_num=0; rx_freq_num<beam_direction_counts.size(); rx_freq_num++) {
+
+      auto num_beams = beam_direction_counts[rx_freq_num];
+
+      // Increment to start of new frequency dataset.
+      auto main_sample_offset = num_samples * (num_main_ants + num_intf_ants) * rx_freq_num;
+      auto main_sample_ptr = filtered_samples.data() + main_sample_offset;
+
+      auto main_phase_ptr = phases.data() + main_phase_offset;
+
+      auto main_results_ptr = beamformed_samples_main.data() + main_results_offset;
+
+      beamform_from_offsets(main_sample_ptr, main_sample_ptr, main_results_ptr, 
+                            num_main_ants, num_beams);
+
+      // Only need to worry about beamforming the interferometer if its being used.
+      if (num_intf_ants > 0) {
+
+        // Skip the main array samples.
+        auto intf_sample_offset = main_sample_offset + (num_samples * num_main_ants);
+        auto intf_sample_ptr = filtered_samples.data() + intf_sample_offset;
+
+        auto intf_phase_offset = main_phase_offset + (num_beams * num_main_ants);
```

Comment:
I think this might be incorrect. sigprocpacket sends for each beam phase delays in order [M0, M1, M2, .... I2, I3] for each beam direction. So to get all the interferometer phase delays for multiple beams together sequentially you would have to pull these out and jump over the main phases. I think this will affect your main phase data above as well. Looks like you are saying your order is [beam 1 M0 ... M15, beam2 M0... M15, ........ beam 1 I0... I3, etc.] Unless I missed where you have rearranged this.

Reply:
I had to rearrange the data for this. I forgot the data was in that format.